### PR TITLE
[sycl-free-function] Refactor Multinomial by free function

### PR DESCRIPTION
**Pull request by jiafuzha**
pytorch  test/test_ops.py -k multinomial  (XPU)                            │ 9 run, OK (7 skipped)  

     pytorch  test/test_torch.py -k multinomial  (CPU)                          │ 16 run, OK             

     xpu-ops  test_torch_xpu.py -k multinomial                                  │ 16 run, OK (2 slow-skipped)    

     xpu-ops slow:  multinomial_invalid_probs ,  multinomial_rng_state_advance  │ 2 run, OK              

     xpu-ops  test_ops_xpu.py -k multinomial                                    │ 17 run, OK (1 expected failure) 

     xpu-ops  extended/test_ops_xpu.py -k multinomial                           │ 16 run, OK             

     xpu-ops  test_distributions_xpu.py -k categorical                          │ 13 run, 1 error 

The one failure,  test_relaxed_one_hot_categorical_2d_xpu , is unrelated to this change: it is a  gradcheck  numerical instability of   
      RelaxedOneHotCategorical.log_prob  at temperature 0.25 (analytic grads ~1e8, numeric diff nan). That path uses Gumbel/softmax and never calls the   multinomial kernel; it fails deterministically across 3 reruns, i.e. a pre-existing XPU issue.

### CODE REVIEW

**Comment by tye1** [comment]
@torchxpubot  check-ut

**Comment by tye1** [comment]
@jiafuzha  pls. fix the code conflicts.

**Comment by jiafuzha** [comment]
@torchxpubot ut-check

File: src/ATen/native/xpu/sycl/MultinomialKernel.cpp [hunk 1]
```diff
@@ -66,35 +66,13 @@ inline void renormRowsL1(
...
-  int64_t cols;
-  scalar_t* t_ptr;
-  int group_size_;
-  sycl_local_acc_t<scalar_t> smem;
-};
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((syclexp::nd_range_kernel<1>))
+void renorm_rows_kernel(int64_t rows, int64_t cols, scalar_t* t_ptr) {
+  auto item = syclext::this_work_item::get_nd_item<1>();
+  unsigned char* smem =
+      (unsigned char*)syclexp::get_work_group_scratch_memory();
```
Comment (gujinghui) [inline src/ATen/native/xpu/sycl/MultinomialKernel.cpp]:
```suggestion
      static_cast<unsigned char*>(syclexp::get_work_group_scratch_memory());
```

File: src/ATen/native/xpu/sycl/MultinomialKernel.cpp [hunk 2]
```diff
@@ -212,219 +195,170 @@ inline void sampleMultinomialWithReplacement(
...
+    int64_t distributions,
+    int categories,
+    const scalar_t* sampled,
+    const scalar_t* dist,
+    int stride_dist,
+    int stride_categories,
+    int group_size) {
+  auto item = syclext::this_work_item::get_nd_item<1>();
+
+  char* scratch = (char*)syclexp::get_work_group_scratch_memory();
```
Comment (gujinghui) [inline src/ATen/native/xpu/sycl/MultinomialKernel.cpp]:
```suggestion
  char* scratch = static_cast<char*>(syclexp::get_work_group_scratch_memory());
```

**Comment by tye1** [comment]
@torchxpubot  check-ut
